be/src/storage/primary_index.cpp

```diff
@@ -746,12 +529,178 @@ class HashIndexImpl<Slice> : public HashIndex {
+        auto& p = _maps[L];                                \
+        if (!p) {                                          \
+            p.reset(new FixSliceHashIndex<(L + 3) / 4>()); \
+        }                                                  \
+        return p.get();                                    \
+    }
+        switch (len) {
+            CASE_LEN(1)
+            CASE_LEN(2)
+            CASE_LEN(3)
+            CASE_LEN(4)
+            CASE_LEN(5)
+            CASE_LEN(6)
+            CASE_LEN(7)
+            CASE_LEN(8)
+            CASE_LEN(9)
+            CASE_LEN(10)
+            CASE_LEN(11)
+            CASE_LEN(12)
+            CASE_LEN(13)
+            CASE_LEN(14)
+            CASE_LEN(15)
+            CASE_LEN(16)
+            CASE_LEN(17)
+            CASE_LEN(18)
+            CASE_LEN(19)
+            CASE_LEN(20)
+            CASE_LEN(21)
+            CASE_LEN(22)
+            CASE_LEN(23)
+            CASE_LEN(24)
+            CASE_LEN(25)
+            CASE_LEN(26)
+            CASE_LEN(27)
+            CASE_LEN(28)
+            CASE_LEN(29)
+            CASE_LEN(30)
+            CASE_LEN(31)
+            CASE_LEN(32)
+            CASE_LEN(33)
+            CASE_LEN(34)
+            CASE_LEN(35)
+            CASE_LEN(36)
+            CASE_LEN(37)
+            CASE_LEN(38)
+            CASE_LEN(39)
+        default: {
+            auto& p = _maps[0];
+            if (!p) p.reset(new SliceHashIndex());
+            return p.get();
+        }
+        }
+    }
+
+public:
+    ShardByLengthSliceHashIndex() = default;
+    ~ShardByLengthSliceHashIndex() override = default;
+
+    size_t size() const override {
+        size_t ret = 0;
+        for (int i = 0; i < max_fix_length; i++) {
+            if (_maps[i]) {
+                ret += _maps[i]->size();
+            }
+        }
+        return ret;
+    }
+
+    size_t capacity() const override {
+        size_t ret = 0;
+        for (int i = 0; i < max_fix_length; i++) {
+            if (_maps[i]) {
+                ret += _maps[i]->capacity();
+            }
+        }
+        return ret;
+    }
+
+    void reserve(size_t size) override {
+        // cannot do reserve for sharding map
+    }
+
+    Status insert(uint32_t rssid, const vector<uint32_t>& rowids, const vectorized::Column& pks, uint32_t idx_begin,
+                  uint32_t idx_end) override {
+        if (idx_begin < idx_end) {
+            auto* keys = reinterpret_cast<const Slice*>(pks.raw_data());
+            for (uint32_t i = idx_begin + 1; i < idx_end; i++) {
+                if (keys[i].size != keys[idx_begin].size) {
+                    RETURN_IF_ERROR(
+                            get_index_by_length(keys[idx_begin].size)->insert(rssid, rowids, pks, idx_begin, i));
+                    idx_begin = i;
+                }
+            }
+            RETURN_IF_ERROR(get_index_by_length(keys[idx_begin].size)->insert(rssid, rowids, pks, idx_begin, idx_end));
+        }
+        return Status::OK();
```

Comment:
return OK directly if begin >= end ? an incorrect caller will get OK from callee.

Reply:
this behavior is consistent with other methods, so let's keep it this way.